cluster/swarm/cluster.go

```diff
@@ -151,32 +151,45 @@ func (c *Cluster) createContainer(config *cluster.ContainerConfig, name string,
 	}
 
 	nodes, err := c.scheduler.SelectNodesForContainer(c.listNodes(), configTemp)
+	c.scheduler.Unlock()
+
 	if err != nil {
-		c.scheduler.Unlock()
 		return nil, err
 	}
-	n := nodes[0]
-	engine, ok := c.engines[n.ID]
-	if !ok {
+
+	for _, n := range nodes {
+		engine, ok := c.engines[n.ID]
+		if !ok {
+			continue
+		}
+
+		c.scheduler.Lock()
+		c.pendingContainers[swarmID] = &pendingContainer{
+			Name:   name,
+			Config: config,
+			Engine: engine,
+		}
 		c.scheduler.Unlock()
-		return nil, fmt.Errorf("error creating container")
-	}
 
-	c.pendingContainers[swarmID] = &pendingContainer{
-		Name:   name,
-		Config: config,
-		Engine: engine,
-	}
+		container, err := engine.Create(config, name, true)
 
-	c.scheduler.Unlock()
+		// pendingContainer need to be clear
+		c.scheduler.Lock()
+		delete(c.pendingContainers, swarmID)
+		c.scheduler.Unlock()
 
-	container, err := engine.Create(config, name, true)
+		if err == nil {
+			return container, err
+		}
+		log.WithFields(log.Fields{"name": "swarm"}).Infof("container create failed: %v", err.Error())
 
-	c.scheduler.Lock()
-	delete(c.pendingContainers, swarmID)
-	c.scheduler.Unlock()
+		if err == dockerclient.ErrConnectionRefused {
```

Comment:
I think there are other classes of errors than `ErrConnectionRefused`.

From what I've seen in production, we might get simple `i/o errors` and so on.

`dockerclient` is able to distinguish between a docker generated error (as in, the error comes from the HTTP **response**) vs a communication error (the error comes from the HTTP client while making the request)